Mathlib/Data/Multiset/Fintype.lean

```diff
@@ -176,59 +174,136 @@ def Multiset.coeEquiv (m : Multiset α) : m ≃ m.toEnumFinset where
+theorem card_toEnumFinset (m : Multiset α) : m.toEnumFinset.card = Multiset.card m := by
   rw [Finset.card, ← Multiset.card_map Prod.fst m.toEnumFinset.val]
   congr
   exact m.map_toEnumFinset_fst
 
 @[simp]
-theorem Multiset.card_coe (m : Multiset α) : Fintype.card m = Multiset.card m := by
+theorem card_coe (m : Multiset α) : Fintype.card m = Multiset.card m := by
   rw [Fintype.card_congr m.coeEquiv]
   simp only [Fintype.card_coe, card_toEnumFinset]
 
 @[to_additive]
-theorem Multiset.prod_eq_prod_coe [CommMonoid α] (m : Multiset α) : m.prod = ∏ x : m, (x : α) := by
+theorem prod_eq_prod_coe [CommMonoid α] (m : Multiset α) : m.prod = ∏ x : m, (x : α) := by
   congr
   simp
 
 @[to_additive]
-theorem Multiset.prod_eq_prod_toEnumFinset [CommMonoid α] (m : Multiset α) :
+theorem prod_eq_prod_toEnumFinset [CommMonoid α] (m : Multiset α) :
     m.prod = ∏ x ∈ m.toEnumFinset, x.1 := by
   congr
   simp
 
 @[to_additive]
-theorem Multiset.prod_toEnumFinset {β : Type*} [CommMonoid β] (m : Multiset α) (f : α → ℕ → β) :
+theorem prod_toEnumFinset {β : Type*} [CommMonoid β] (m : Multiset α) (f : α → ℕ → β) :
     ∏ x ∈ m.toEnumFinset, f x.1 x.2 = ∏ x : m, f x x.2 := by
   rw [Fintype.prod_equiv m.coeEquiv (fun x ↦ f x x.2) fun x ↦ f x.1.1 x.1.2]
   · rw [← m.toEnumFinset.prod_coe_sort fun x ↦ f x.1 x.2]
   · intro x
     rfl
+
+/--
+If `s = t` then there's an equivalence between the appropriate types.
+-/
+@[simps]
+def cast {s t : Multiset α} (h : s = t) : s ≃ t where
+  toFun x := ⟨x.1, x.2.cast (by simp [h])⟩
+  invFun x := ⟨x.1, x.2.cast (by simp [h])⟩
+  left_inv x := rfl
+  right_inv x := rfl
+
+instance : IsEmpty (0 : Multiset α) := Fintype.card_eq_zero_iff.mp (by simp)
+
+instance : IsEmpty (∅ : Multiset α) := Fintype.card_eq_zero_iff.mp (by simp)
+
+/--
+`v ::ₘ m` is equivalent to `Option m` by mapping one `v` to `none` and everything else to `m`.
+-/
+def consEquiv {v : α} : v ::ₘ m ≃ Option m where
+  toFun x := if h : x.1 = v ∧ x.2.val = m.count v then none else some ⟨x.1, ⟨x.2, by
+    by_cases hv : x.1 = v
+    · simp only [hv, true_and] at h ⊢
+      apply lt_of_le_of_ne (Nat.le_of_lt_add_one _) h
+      convert x.2.2 using 1
+      simp [hv]
+    · convert x.2.2 using 1
+      exact (count_cons_of_ne hv _).symm
+    ⟩⟩
+  invFun x := x.elim ⟨v, ⟨m.count v, by simp⟩⟩ (fun x ↦ ⟨x.1, x.2.castLE (count_le_count_cons ..)⟩)
+  left_inv := by
+    rintro ⟨x, hx⟩
+    dsimp only
+    split
+    · rename_i h
+      obtain ⟨rfl, h2⟩ := h
+      simp [← h2]
+    · simp
+  right_inv := by
+    rintro (_ | x)
+    · simp
+    · simp only [Option.elim_some, Nat.zero_eq, Fin.coe_castLE, Fin.eta, Sigma.eta, dite_eq_ite,
+      ite_eq_right_iff, reduceCtorEq, imp_false, not_and]
+      apply ne_of_lt ∘ (· ▸ x.2.2)
+
+@[simp]
+lemma consEquiv_symm_none {v : α} :
+    (consEquiv (m := m) (v := v)).symm none =
+      ⟨v, ⟨m.count v, (count_cons_self v m) ▸ (Nat.lt_add_one _)⟩⟩ :=
+  rfl
```

Comment:
We can state the `symm_some` lemma too, right?

For that matter the forward direction for `x = v` (in two cases) as well.